Extend try...except translation support

devanlai opened the pull request:
Preliminary pull request to add support for more forms of try...except.
The following forms are now accepted:
Exceptions from other modules:

    try:
        value = struct.unpack("foo")
    except struct.error:
        print("oops")
    ---
    import * as struct from 'struct';
    var value;
    try {
        value = struct.unpack("foo");
    } catch(e) {
        if ((e instanceof struct.error)) {
            console.log("oops");
        } else {
            throw e;
        }
    }

Matching against a sequence of exception types (when known at transpile-time):

    try:
        value = 1/input("enter zero here")
    except (KeyboardInterrupt, ZeroDivisionError):
        print("oops")
    ---
    var value;
    try {
        value = (1 / input("enter zero here"));
    } catch(e) {
        if (((e instanceof KeyboardInterrupt) || (e instanceof ZeroDivisionError))) {
            console.log("oops");
        } else {
            throw e;
        }
    }


Using different variable names in each except clause:

    try:
        1/0
    except ZeroDivisionError as err1:
        print("oops", err1)
    except RuntimeError as err2:
        print("uhoh", err2)
    ---
    try {
        (1 / 0);
    } catch(err2) {
        if ((err2 instanceof ZeroDivisionError)) {
            let err1 = err2;
            console.log("oops", err1);
        } else {
            if ((err2 instanceof RuntimeError)) {
                console.log("uhoh", err2);
            } else {
                throw err2;
            }
        }
    }

Open questions:
To handle sequences of exceptions, I borrowed some of the code from `Call_isinstance` and split it out into a separate function, `_build_call_isinstance`. Do you have any thoughts on where that functionality should live? Right now I'm just directly importing it from the classes module, which is terrible.

Is there a established way to update the expected test output? I could find any way to transpile the code with body_only=True short of calling `translate_object` directly.

azazel75 commented:
Thanks for doing this, are you actually using this tool?  I ask you just because you are one of those few people that contributed some actual code and I'm curious about what you do with it...

By the way, could you please move that `_build_call_isinnstance()` to a `common.py` module and import from there?

Another thing, I see that your code uses a `let` declaration, and unfortunately I must say that I don't like it so much, for the following reasons:

* let has no correspondent in the python world and the code of the transpiler doesn't yet have a so fine grained way of dealing blocks of code to be able to deal with variable's usage boundaries smaller than functions, and that's why it's not actually used in the code, the same and more can be said for `const`;

* `let` depends on ES6 and I would like for the  try...catch statement to be available in a simpler ES5-compatible form.

For these reasons, I would like to ask you to add the custom symbol name to the ones that will end up in the `var` declaration...

azazel75 commented:
oh, nevermind, you can trash all the declaration step maybe.. from the mozilla JS ref:
```
...  The catch block is unique in that JavaScript creates this identifier when the catch block is entered and it adds it to the current scope; the identifier lasts only for the duration of the catch block; after the catch block finishes executing, the identifier is no longer available.
```

azazel75 commented:
It does the "right thing" already with one exception type:
```python
$ pj -s -
try:
        a=2
except KeyError as pollo:
        print('ciao')
```
```javascript
var a;
try {
    a = 2;
} catch(pollo) {
    if ((pollo instanceof KeyError)) {
        console.log("ciao");
    } else {
        throw pollo;
    }
}```

devanlai commented:
Hi @azazel75,
I'm playing around with javascripthon to speed up porting some Python code to javascript by generating some code that has the right business logic and then fixing up the places where APIs mismatch.

I've moved `_build_call_isinstance` to a new `common.py` module and switched from `let` to `var`.
The rename is only necessary when using two handlers with different variable names for the exception in each block. It's not a common situation, but I figured I would handle that case while I was poking around the code anyways.

azazel75 commented:
Well done, thanks!
